openhands-tools/openhands/tools/delegate/impl.py

```diff
@@ -0,0 +1,405 @@
+                f"event_type={type(event).__name__}, "
+                f"source={getattr(event, 'source', 'N/A')}"
+            )
+
+            if not (isinstance(event, MessageEvent) and event.source == "agent"):
+                logger.debug(
+                    f"Sub-agent {sub_conversation_id[:8]} ignoring non-agent "
+                    f"message event"
+                )
+                return
+
+            if not (hasattr(event, "llm_message") and event.llm_message):
+                logger.debug(
+                    f"Sub-agent {sub_conversation_id[:8]} event has no llm_message"
+                )
+                return
+
+            message_parts = []
+            for content in event.llm_message.content:
+                if isinstance(content, TextContent):
+                    message_parts.append(content.text)
+                else:
+                    message_parts.append(f"[{type(content).__name__}]")
+
+            message_text = " ".join(message_parts)
+
+            if not message_text.strip():
+                logger.debug(
+                    f"Sub-agent {sub_conversation_id[:8]} sent empty message, ignoring"
+                )
+                return
+
+            parent_message = f"[Sub-agent {sub_conversation_id[:8]}]: {message_text}"
+            logger.info(
+                f"🔄 Sub-agent {sub_conversation_id[:8]} sending "
+                f"message to parent: {message_text[:100]}..."
+            )
+
+            message = Message(
+                role="user",
+                content=[TextContent(text=parent_message)],
+            )
+            try:
+                self._pending_parent_messages.put_nowait(message)
+                queue_size = self._pending_parent_messages.qsize()
+                logger.info(
+                    f"✅ Queued message from sub-agent "
+                    f"{sub_conversation_id[:8]} "
+                    f"(queue size: {queue_size})"
+                )
+                self._trigger_parent_if_idle()
+            except queue.Full:
+                logger.warning(
+                    f"❌ Parent message queue is full, dropping "
+                    f"message from sub-agent {sub_conversation_id[:8]}"
+                )
+
+        return callback
+
+    def _run_sub_agent(
+        self,
+        sub_conversation: "LocalConversation",
+        initial_message: str,
+    ):
+        """Run a sub-agent in a separate thread.
+
+        Args:
+            sub_conversation_id: The ID of the sub-agent conversation
+            sub_conversation: The conversation object for the sub-agent
+            initial_message: The initial message to send to the sub-agent
+        """
+        sub_conversation_id = str(sub_conversation.id)
+        try:
+            with self._lock:
+                if sub_conversation_id in self._sub_agents:
+                    self._sub_agents[sub_conversation_id].state = SubAgentState.RUNNING
```

Comment:
I wonder if we should just maintain `Converstaion` objects and `Conversation` should be able to manage its own state?